### ISSUE TITLE
Request to add .shape and .ndim attributes to TensorStore objects

### ISSUE THREAD

**Issue by jni** [google/tensorstore#17]
Thank you @jbms for answering my [StackOverflow question](https://stackoverflow.com/questions/64924224/getting-a-view-of-a-zarr-array-slice/64962223#64962223)!

For a bit of context, the reason @AbigailMcGovern and I need this is to use [napari](https://napari.org) to paint segmentations into (parts of) .zarr files. Now we are able to use tensorstore (🎉🙏), *but only* after patching in `.shape` and `.ndim` attributes, which napari expects of its inputs at various points. These are quite standard in the Python array world, with NumPy, dask, zarr, PyTorch, and Tensorflow all supporting them, among others.

I understand API creep is real but I hope that these are standard enough to be considered for inclusion in tensorstore!

Thank you!

**Comment by jbms** [comment]
Adding in those attributes is no problem.   The only question is whether shape needs to be returned as a tuple or if it can be returned as a 1-d numpy array, like .domain.shape currently?

**Comment by jbms** [comment]
Are other attributes like `size` or `T` also important?  They would of course also be easy to add.

The biggest incompatibility with TensorStore is that it supports a non-zero origin for dimensions, and the origin can be negative as well.  If you don't use a non-zero origin you can mostly ignore it, but it does mean that negative indexing behaves differently in TensorStore than in numpy.

**Comment by jni** [comment]
> The only question is whether shape needs to be returned as a tuple or if it can be returned as a 1-d numpy array, like .domain.shape currently?

> Are other attributes like size or T also important? They would of course also be easy to add.

These are good questions! I was waffling my way through them when I remembered this blog post:

https://data-apis.org/blog/array_api_standard_release/

So, these issues are being worked out as we speak and the current API standard is in draft:

https://data-apis.github.io/array-api/latest/API_specification/index.html

You could definitely add your voice to that crowd! The repo is https://github.com/data-apis/array-api. But, more concretely, your questions are indeed answered in the current spec:

- yes, shape should be a tuple, *or* a "shape object", but I can't find what a shape object is... Tuple seems "safe".
- yes, .size and .T are part of the standard API.
- so is .dtype
- but that's all the attributes, which is nice!

**Comment by jbms** [comment]
Thanks for your feedback.

Yes, I had seen:

https://data-apis.github.io/array-api/latest/API_specification/index.html

which is why I suggested those attributes.

Regarding the shape: I find it is generally more convenient to work with shapes as numpy arrays because I often want to do vector (e.g. component-wise) arithmetic operations on them.  Currently `domain.shape` returns a numpy array with the NPY_ARRAY_WRITEABLE flag cleared to make it non-writeable.  I had initially assumed that did indeed make it immutable (as long as the user doesn't manually reset the writeable the flag, which I think need not be considered).  However, thinking about it more in response to your comment, I realized the user can still call `resize` on the returned shape array, and there doesn't appear to be any way to prevent that.  That doesn't cause any problems to TensorStore itself --- every access to the `shape` property just returns a fresh ndarray, so as long as the user doesn't set the writable flag and modify the array data itself, TensorStore's own state remains consistent.  However, if the user shares references to the returned shape vector, and expects it to be immutable, but some other code calls resize on it, then that could be a problem.  Still, it seems like calling resize may be sufficiently unusual that the returned shape object could still be considered "immutable".

I could alternatively change to always return a tuple, which would be more consistent with other libraries, but perhaps less convenient for when you want to do arithmetic operations on the shape vector.

Do you know for your own code (i.e. napari) whether you count on the shape being a tuple?

**Comment by jni** [comment]
> Still, it seems like calling resize may be sufficiently unusual that the returned shape object could still be considered "immutable".

I agree with this statement.

> Do you know for your own code (i.e. napari) whether you count on the shape being a tuple?

I don't know, but I think @AbigailMcGovern could test this very quickly. Either way, if we do that's kinda bad behaviour from us so I'd be happy to modify. Personally I like the arguments you made above so I'm leaning towards your sort-of-almost-immutable arrays. 👍 

> perhaps less convenient for when you want to do arithmetic operations on the shape vector.

Yes, we do often do `np.array(image.shape)`, so again, points for the array-from-the-get-go approach.

**Comment by jni** [comment]
Awesome! Thank you @jbms! 🙏 

Quick usage question! We've noticed that `arr[start:stop]` returns a view that is still indexed by values between start and stop, in contrast to 0:stop-start as you would find in NumPy & co. This is a great feature but napari does not like it. =P Is there a way for us to keep that same view but transform the indexing?

**Comment by jbms** [comment]
`arr[ts.d[:][start:stop].translate_to[0]]` or `arr[ts.d[0][start:stop].translate_to[0]]` depending on whether `start` and `stop` are scalars or vectors.

Or for an existing view, you can just use: `arr[ts.d[:].translate_to[0]]`.

It would probably make sense to add `translate_to` (and other DimExpression methods) as a method of `tensorstore.TensorStore` directly as well, but for now it can only be used on a DimExpression.

Breaking numpy compatibility with the indexing here was rather unfortunate, but there could only be one meaning for `[...]` indexing and preserving the origin seemed like the only sane option given that non-zero origins are allowed.  TensorStore itself does automatically align origins for some operations (https://google.github.io/tensorstore/index_space.html#alignment-and-broadcasting).  I'm certainly open to suggestions on how to better handle this.

**Comment by jni** [comment]
> Breaking numpy compatibility with the indexing here was rather unfortunate, but there could only be one meaning for `[...]` indexing and preserving the origin seemed like the only sane option given that non-zero origins are allowed. TensorStore itself does automatically align origins for some operations (google.github.io/tensorstore/index_space.html#alignment-and-broadcasting). I'm certainly open to suggestions on how to better handle this.

Yeah it's tricky and I agree with everything you said. It's unclear to me that the NumPy model is better and the `translate_to` workaround should suit us fine, so I wouldn't rush to touch it for now. [PEP-637](https://www.python.org/dev/peps/pep-0637/) should provide a nifty solution if it gets accepted.

Having said that `ts.d[:].translate_to[0]` with the multiple getitems reads a bit weird to me. I might suggest something like `ts.Indexer(translate_to=0)`? Also, just to make sure I fully grok that expression, if instead I'd created by view with `arr[:, start:stop]`, would the corresponding expression be `arr[:, ts.d[:].translate_to[0]]`?

**Comment by jbms** [comment]
`ts.d` starts a DimExpression (https://google.github.io/tensorstore/python/indexing.html#dimension-expressions) which allows you to perform a sequence of indexing operations to all of a subset of the dimensions.

`ts.d[:]` selects all dimensions, and `ts.d[1, 0]` selects dimensions 1 and 0, and `ts.d["x", "y"]` selects dimensions by name.  You can then chain a sequence of indexing operations that apply just to those dimensions.  `ts.d[:].translate_to[0]` translates the origin of all dimensions to 0.  The argument to translate_to can either be a scalar, in which case it used for all selected dimensions, or a sequence of numbers, in which case the length must equal the number of selected dimensions.  `arr[ts.d[1].label["x"][5:10].translate_to[0]]` returns a view of `arr` where we have restricted the domain of dimension 1 to 5:10, translated what was 5 to become 0, and labeled it "x".

You can't use a DimExpression as part of another numpy-style indexing  operation, as in `arr[:, ts.d[:].translate_to[0]]`, but you can chain them since they return views.  The DimExpression must be the only argument to __getitem__.  I suppose mixing could be supported if consistent semantics could be defined.  I do see that the use of getitem for both applying a DimExpression and regular numpy indexing is a bit confusing, but it made some sense to me since it is still indexing and __getitem__ suggests indexing, and it also keeps the DimExpression syntax concise.
